library/x509_create.c

```diff
@@ -86,62 +92,141 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+                    d = data;
+                    if( *s == '\"' )
+                    {
+                        dn_cur_state = dn_no_tag_in_quot;
+                        s++;
+                        c += 2;
+                    }
+                    else
+                    {
+                        dn_cur_state = dn_no_tag_no_quot;
+                    }
+                }
             }
+            break;
 
-            s = c + 1;
-            in_tag = 0;
-            d = data;
-        }
-
-        if( !in_tag && *c == '\\' && c != end )
-        {
-            c++;
-
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            case dn_no_tag_in_quot:
             {
-                ret = MBEDTLS_ERR_X509_INVALID_NAME;
-                goto exit;
+                if( *c == '\"' && c != end )
+                {
+                    c++;
+                    s = c + 1;
+                    if( *c != ',' )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                    ret = dn_store_named_data( head, oid, data, d, &s,
+                                               &dn_cur_state, &c, end );
+                    if( ret != 0 )
+                    {
+                        goto exit;
+                    }
+                }
             }
-        }
-        else if( !in_tag && ( *c == ',' || c == end ) )
-        {
-            if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
-                                       (unsigned char *) data,
-                                       d - data ) == NULL )
+            break;
+
+            case dn_no_tag_no_quot:
             {
-                return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+                if( *c == ',' || c == end )
+                {
+                    if( d - data == 0 )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                    ret = dn_store_named_data( head, oid, data, d, &s,
+                                               &dn_cur_state, &c, end );
+                    if( ret != 0 )
+                    {
+                        goto exit;
+                    }
+                }
+                if( *c == '\\' && c != end )
+                {
+                    c++;
+
+                    ret = x509_dn_check_special_chars( &c, end );
+                    if( ret != 0 )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                }
             }
-
-            while( c < end && *(c + 1) == ' ' )
-                c++;
-
-            s = c + 1;
-            in_tag = 1;
         }
 
-        if( !in_tag && s != c + 1 )
+        if( ( dn_cur_state != dn_in_tag_no_quot ) &&
+            ( s != c + 1 ) )
```

Comment:
To my understanding, this is supposed to copy any character within a DN value into the temporary `data` buffer. However, it doesn't work for e.g. for a DN name `A=""`, in which case the correct behaviour would be to parse an empty quoted string (and error out); instead, the code would read the second `"` into the `data` buffer.

I suggest to move the copying code into the sections handling the states `dn_in_quoted_val` and `dn_in_unquoted_value`, once it is known that the next character does not lead to a state change. This way, the logic is very simple: For any state, check if the current character is state-changing; if yes, change the state; if not, potentially unescape it and copy it to the target buffer (a no-op for the `s` buffer, and an actual copy for the `d` buffer).

Reply:
I have added a test case for this(not committed yet), and we return a failure.

Reply:
Yes, but the internal treatment is incorrect nonetheless; we should never copy a delimiting quote into the DN value buffer.

Reply:
Yes, I see your point. I'll fix this

Reply:
I find this still hard to read. Suggestion: Add a local variable `copy_cur` or similar which is set by the state handling for `dn_in_quoted_value` and `dn_in_unquoted_value` if they found a character that should be copied to the value buffer `data`, and use that here.

Reply:
I don't quite understand what you mean with your last comment